Navigation fails when using Goto Implementation / Type Declaration

A navigation issue is observed when using `Goto → Implementation` and `Goto → Type Declaration` actions.
This problem is introduced in LSP4IJ `0.19.0` and does not occur in previous versions.

**Steps to Reproduce**
1. Open any Java class.
2. Use `Cmd/Ctrl + Click` on a class name (e.g., String) to navigate to its implementation.
3. The issue occurs.

**Alternative reproduction paths:**
1. Use `Goto → Implementation.`
2. Use `Goto → Type Declaration` on a variable (e.g., `String str = "";`)
3. Select str → `Goto → Type Declaration`.


```
Dec 05, 2025 6:58:53 PM org.eclipse.lsp4mp.ls.MicroProfileLanguageServer initialize
INFO: Initializing MicroProfile server 0.13.0-20241022-1138 with /Applications/IntelliJ IDEA CE.app/Contents/jbr/Contents/Home
Dec 05, 2025 7:05:02 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4mp.ls.commons.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:148)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:115)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4mp.ls.java.JavaTextDocuments$JavaTextDocument.executeIfInMicroProfileProject(java.util.function.BiFunction, Object, boolean)" because "document" is null
	at org.eclipse.lsp4mp.ls.java.JavaFileTextDocumentService.codeLens(JavaFileTextDocumentService.java:203)
	at org.eclipse.lsp4mp.ls.MicroProfileTextDocumentService.codeLens(MicroProfileTextDocumentService.java:242)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 14 more
```

<img width="2672" height="846" alt="Image" src="https://github.com/user-attachments/assets/bd20a32c-0806-4024-8a34-e75a241ac2d6" />

--- a/src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java
@@ -43,6 +43,7 @@ import com.intellij.psi.PsiManager;
 import com.redhat.devtools.lsp4ij.client.features.FileUriSupport;
 import com.redhat.devtools.lsp4ij.internal.SimpleLanguageUtils;
 import com.redhat.devtools.lsp4ij.internal.StringUtils;
+import com.redhat.devtools.lsp4ij.internal.uri.UriConverterManager;
 import com.redhat.devtools.lsp4ij.usages.LocationData;
 import org.apache.commons.io.FileUtils;
 import org.eclipse.lsp4j.*;
@@ -67,14 +68,6 @@ import java.util.concurrent.atomic.AtomicBoolean;
 public class LSPIJUtils {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(LSPIJUtils.class);
-
-    public static final String JAR_PROTOCOL = "jar";
-
-    public static final String JRT_PROTOCOL = "jrt";
-
-    private static final String JAR_SCHEME = JAR_PROTOCOL + ":";
-
-    private static final String JRT_SCHEME = JRT_PROTOCOL + ":";
 
     public static final String HASH_SEPARATOR = "#";
 
@@ -490,26 +483,7 @@ public class LSPIJUtils {
     }
 
     public static @NotNull URI toUri(@NotNull File file) {
-        String path = file.getPath();
-        if (path.startsWith("\\\\")) {
-            // Extract authority and path from Windows UNC path: \\authority\path\…
-            String uncPath = path.substring(2); // Remove leading \\
-            int firstSep = uncPath.indexOf('\\');
-            if (firstSep > 0) {
-                String authority = uncPath.substring(0, firstSep);
-                String uriPath = uncPath.substring(firstSep).replace('\\', '/');
-                // Use lenient parsing to support WSL authority 'wsl$'
-                return URI.create("file://" + authority + uriPath);
-            }
-        }
-        
-        // URI scheme specified by language server protocol and LSP
-        try {
-            return new URI("file", "", file.getAbsoluteFile().toURI().getPath(), null); //$NON-NLS-1$ //$NON-NLS-2$
-        } catch (URISyntaxException e) {
-            LOGGER.warn(e.getLocalizedMessage(), e);
-            return file.getAbsoluteFile().toURI();
-        }
+        return UriConverterManager.getInstance().toUri(file);
     }
 
     public static @Nullable URI toUri(@NotNull PsiFile psiFile) {
@@ -518,7 +492,7 @@ public class LSPIJUtils {
     }
 
     public static @NotNull URI toUri(@NotNull VirtualFile file) {
-        return toUri(VfsUtilCore.virtualToIoFile(file));
+        return UriConverterManager.getInstance().toUri(file);
     }
 
     public static @Nullable String toUriAsString(@NotNull PsiFile psFile) {
@@ -1076,10 +1050,6 @@ public class LSPIJUtils {
         return findResourceFor(toUri(file).toString());
     }
 
-    public static @Nullable VirtualFile findResourceFor(@NotNull URI uri) {
-        return LocalFileSystem.getInstance().findFileByIoFile(Paths.get(uri).toFile());
-    }
-
     /**
      * Returns the virtual file from the given uri and null otherwise.
      *
@@ -1087,23 +1057,7 @@ public class LSPIJUtils {
      * @return the virtual file from the given uri and null otherwise.
      */
     public static @Nullable VirtualFile findResourceFor(@NotNull String uri) {
-        if (uri.startsWith(JAR_SCHEME) || uri.startsWith(JRT_SCHEME)) {
-            // ex : jar:file:///C:/Users/azerr/.m2/repository/io/quarkus/quarkus-core/3.0.1.Final/quarkus-core-3.0.1.Final.jar!/io/quarkus/runtime/ApplicationConfig.class
-            try {
-                return VfsUtil.findFileByURL(new URL(uri));
-            } catch (MalformedURLException e) {
-                return null;
-            }
-        }
-        if (uri.contains("%")) {
-            // ex : file:///c%3A/Users/azerr/IdeaProjects/untitled7/test.js
-            // the uri must be decoded (ex : file:///c:/Users) otherwise IntelliJ cannot retrieve the virtual file.
-            // Keep the original '+' after decoding the Uri
-            uri = uri.replace("+", "%2B");
-            // Decode the uri
-            uri = URLDecoder.decode(uri, StandardCharsets.UTF_8);
-        }
-        return VirtualFileManager.getInstance().findFileByUrl(VfsUtilCore.fixURLforIDEA(uri));
+        return UriConverterManager.getInstance().findResourceFor(uri);
     }
 
     /**

--- a/src/main/java/com/redhat/devtools/lsp4ij/client/features/FileUriSupportBase.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/client/features/FileUriSupportBase.java
@@ -10,18 +10,14 @@
  ******************************************************************************/
 package com.redhat.devtools.lsp4ij.client.features;
 
-import com.intellij.openapi.vfs.VfsUtilCore;
 import com.intellij.openapi.vfs.VirtualFile;
 import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.internal.uri.UriConverterManager;
 import com.redhat.devtools.lsp4ij.internal.uri.UriFormatter;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.net.URI;
-import java.util.Objects;
-
-import static com.redhat.devtools.lsp4ij.LSPIJUtils.JAR_PROTOCOL;
-import static com.redhat.devtools.lsp4ij.LSPIJUtils.JRT_PROTOCOL;
 
 /**
  * Provides a base implementation of the {@link FileUriSupport} interface
@@ -45,9 +41,9 @@ public class FileUriSupportBase implements FileUriSupport {
 
     @Override
     public String toString(@NotNull VirtualFile file) {
-        String protocol = file.getFileSystem().getProtocol();
-        if (JAR_PROTOCOL.equals(protocol) || JRT_PROTOCOL.equals(protocol)) {
-            return Objects.requireNonNull(VfsUtilCore.convertToURL(file.getUrl())).toExternalForm();
+        String uriString = UriConverterManager.getInstance().toString(file);
+        if (uriString != null) {
+            return uriString;
         }
         URI fileUri = getFileUri(file);
         return fileUri != null ? toString(fileUri, file.isDirectory()) : null;

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/DefaultUriConverter.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/DefaultUriConverter.java
@@ -1,0 +1,99 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.internal.uri;
+
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * Default {@link UriConverter} implementation for standard {@code file://} URIs.
+ *
+ * <p>
+ * This converter is used for regular files located on the local file system.
+ * It follows the Language Server Protocol (LSP) specification for file URIs
+ * and ensures compatibility with IntelliJ's Virtual File System (VFS).
+ * </p>
+ */
+class DefaultUriConverter implements UriConverter {
+
+    /**
+     * Resolves a {@link VirtualFile} from a {@code file://} URI string.
+     *
+     * <p>
+     * The URI is first normalized using {@link VfsUtilCore#fixURLforIDEA(String)}
+     * to handle platform-specific issues (e.g. Windows paths),
+     * then resolved through IntelliJ's {@link VirtualFileManager}.
+     * </p>
+     *
+     * @param uri the file URI string
+     * @return the corresponding {@link VirtualFile}, or {@code null} if it cannot be resolved
+     */
+    @Override
+    public @Nullable VirtualFile findResourceFor(@NotNull String uri) {
+        return VirtualFileManager.getInstance()
+                .findFileByUrl(VfsUtilCore.fixURLforIDEA(uri));
+    }
+
+    /**
+     * Converts a Java {@link File} to an LSP-compliant {@link URI}.
+     *
+     * <p>
+     * The Language Server Protocol requires {@code file://} URIs with
+     * properly encoded paths and without host components.
+     * </p>
+     *
+     * <p>
+     * This implementation explicitly builds the URI to avoid
+     * platform-specific issues, especially on Windows.
+     * </p>
+     *
+     * @param file the Java file
+     * @return the corresponding {@link URI}
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull File file) {
+        try {
+            // LSP-compliant file URI: file:///absolute/path
+            return new URI(
+                    "file",
+                    "", // no authority
+                    file.getAbsoluteFile().toURI().getPath(),
+                    null
+            );
+        } catch (URISyntaxException e) {
+            // Fallback to the standard Java URI representation
+            return file.getAbsoluteFile().toURI();
+        }
+    }
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} to its {@code file://} URI string representation.
+     *
+     * <p>
+     * This method is intentionally left unimplemented here and may be
+     * provided by a higher-level or composite {@link UriConverter}.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return {@code null}
+     */
+    @Override
+    public @Nullable String toString(@NotNull VirtualFile file) {
+        return null;
+    }
+}

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/JarUriConverter.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/JarUriConverter.java
@@ -1,0 +1,154 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.internal.uri;
+
+import com.intellij.openapi.vfs.VfsUtil;
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFile;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Objects;
+
+/**
+ * {@link UriConverter} implementation dedicated to {@code jar:} and {@code jrt:} URIs.
+ *
+ * <p>
+ * This converter is responsible for handling resources located inside:
+ * <ul>
+ *   <li>JAR files ({@code jar:file://...!/path})</li>
+ *   <li>Java runtime images ({@code jrt:/modules/...})</li>
+ * </ul>
+ *
+ * <p>
+ * These URI types are commonly used by language servers when referring to:
+ * <ul>
+ *   <li>dependencies from Maven/Gradle caches</li>
+ *   <li>JDK classes (via the {@code jrt} file system)</li>
+ * </ul>
+ *
+ * <p>
+ * This converter does <strong>not</strong> handle regular {@code file://} URIs.
+ * </p>
+ */
+class JarUriConverter implements UriConverter {
+
+    /** IntelliJ file system protocol for JAR files */
+    public static final String JAR_PROTOCOL = "jar";
+
+    /** IntelliJ file system protocol for Java runtime image (JRT) */
+    public static final String JRT_PROTOCOL = "jrt";
+
+    /** URI scheme prefix for JAR resources */
+    private static final String JAR_SCHEME = JAR_PROTOCOL + ":";
+
+    /** URI scheme prefix for JRT resources */
+    private static final String JRT_SCHEME = JRT_PROTOCOL + ":";
+
+    /**
+     * Resolves a {@link VirtualFile} from a {@code jar:} or {@code jrt:} URI string.
+     *
+     * <p>
+     * Example:
+     * <pre>
+     * jar:file:///C:/Users/.m2/repository/.../artifact.jar!/com/example/Foo.class
+     * </pre>
+     * </p>
+     *
+     * @param uri the URI string to resolve
+     * @return the corresponding {@link VirtualFile}, or {@code null} if the URI
+     *         is not a JAR/JRT URI or cannot be resolved
+     */
+    @Override
+    public @Nullable VirtualFile findResourceFor(@NotNull String uri) {
+        if (uri.startsWith(JAR_SCHEME) || uri.startsWith(JRT_SCHEME)) {
+            try {
+                return VfsUtil.findFileByURL(new URL(uri));
+            } catch (MalformedURLException e) {
+                // Invalid URI format → resource cannot be resolved
+                return null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts a JAR or JRT {@link VirtualFile} into a {@link URI}.
+     *
+     * <p>
+     * The {@link VirtualFile#getUrl()} value is converted into a {@link URL}
+     * and then into a {@link URI}.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the corresponding {@link URI}, or {@code null} if the file
+     *         does not belong to a JAR or JRT file system
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull VirtualFile file) {
+        String protocol = file.getFileSystem().getProtocol();
+        if (JAR_PROTOCOL.equals(protocol) || JRT_PROTOCOL.equals(protocol)) {
+            try {
+                return Objects
+                        .requireNonNull(VfsUtilCore.convertToURL(file.getUrl()))
+                        .toURI();
+            } catch (URISyntaxException e) {
+                // Ignore invalid URI conversion
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts a {@link File} into a {@link URI}.
+     *
+     * <p>
+     * This implementation does not support conversion from {@link File}
+     * because JAR and JRT resources are not directly backed by {@link File}
+     * instances.
+     * </p>
+     *
+     * @param file the Java file
+     * @return always {@code null}
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull File file) {
+        return null;
+    }
+
+    /**
+     * Converts a JAR or JRT {@link VirtualFile} into its URI string representation.
+     *
+     * <p>
+     * The returned string is suitable for use in Language Server Protocol (LSP)
+     * messages.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the URI string, or {@code null} if the file does not belong to
+     *         a JAR or JRT file system
+     */
+    @Override
+    public @Nullable String toString(@NotNull VirtualFile file) {
+        String protocol = file.getFileSystem().getProtocol();
+        if (JAR_PROTOCOL.equals(protocol) || JRT_PROTOCOL.equals(protocol)) {
+            return Objects
+                    .requireNonNull(VfsUtilCore.convertToURL(file.getUrl()))
+                    .toExternalForm();
+        }
+        return null;
+    }
+}

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/UriConverter.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/UriConverter.java
@@ -1,0 +1,107 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.internal.uri;
+
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFile;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.net.URI;
+
+/**
+ * Responsible for converting between:
+ * <ul>
+ *   <li>LSP-style URIs (file://, jar://, etc.)</li>
+ *   <li>IntelliJ {@link VirtualFile}</li>
+ *   <li>Java {@link File}</li>
+ * </ul>
+ *
+ * <p>
+ * This abstraction centralizes all URI ↔ file conversions needed by LSP4IJ,
+ * including resolution of {@link VirtualFile} instances from URI strings
+ * received from language servers.
+ * </p>
+ *
+ * <p>
+ * This interface is internal to LSP4IJ and is not intended for public API use.
+ * </p>
+ */
+@ApiStatus.Internal
+public interface UriConverter {
+
+    /**
+     * Resolves an IntelliJ {@link VirtualFile} from a URI string.
+     *
+     * <p>
+     * The URI is typically an LSP URI (e.g. {@code file:///path/to/file},
+     * {@code jar://...}, etc.).
+     * </p>
+     *
+     * @param uri the URI string to resolve
+     * @return the corresponding {@link VirtualFile}, or {@code null} if
+     *         the resource cannot be resolved
+     */
+    @Nullable
+    VirtualFile findResourceFor(@NotNull String uri);
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} to a {@link URI}.
+     *
+     * <p>
+     * This default implementation first converts the {@link VirtualFile}
+     * to a Java {@link File} using
+     * {@link VfsUtilCore#virtualToIoFile(VirtualFile)}, then delegates
+     * to {@link #toUri(File)}.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the corresponding {@link URI}, or {@code null} if the conversion fails
+     */
+    default @Nullable URI toUri(@NotNull VirtualFile file) {
+        return toUri(VfsUtilCore.virtualToIoFile(file));
+    }
+
+    /**
+     * Converts a Java {@link File} to a {@link URI}.
+     *
+     * <p>
+     * Implementations may handle special cases such as:
+     * <ul>
+     *   <li>non-local file systems</li>
+     *   <li>case normalization (important on Windows)</li>
+     *   <li>symbolic links</li>
+     * </ul>
+     * </p>
+     *
+     * @param file the Java file
+     * @return the corresponding {@link URI}, or {@code null} if it cannot be created
+     */
+    @Nullable
+    URI toUri(@NotNull File file);
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} to its URI string representation.
+     *
+     * <p>
+     * The returned string is expected to be compatible with the Language Server
+     * Protocol (LSP), typically using {@code file://} URIs.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the URI string, or {@code null} if the conversion fails
+     */
+    @Nullable
+    String toString(@NotNull VirtualFile file);
+
+}

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/UriConverterManager.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/UriConverterManager.java
@@ -1,0 +1,192 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.internal.uri;
+
+import com.intellij.openapi.vfs.VirtualFile;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.net.URI;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+/**
+ * Central {@link UriConverter} dispatcher used by LSP4IJ.
+ *
+ * <p>
+ * This manager delegates URI conversion operations to a chain of
+ * specialized {@link UriConverter} implementations, selecting the
+ * first one able to handle the request.
+ * </p>
+ *
+ * <p>
+ * It follows a <strong>chain-of-responsibility</strong> pattern, allowing
+ * new URI types to be supported without modifying existing converters.
+ * </p>
+ *
+ * <p>
+ * This class is internal to LSP4IJ and exposed as a singleton.
+ * </p>
+ */
+@ApiStatus.Internal
+public class UriConverterManager implements UriConverter {
+
+    /** Singleton instance */
+    private static final UriConverterManager INSTANCE;
+
+    static {
+        INSTANCE = new UriConverterManager();
+    }
+
+    /**
+     * Returns the singleton instance of the URI converter manager.
+     *
+     * @return the shared {@link UriConverterManager} instance
+     */
+    public static UriConverterManager getInstance() {
+        return INSTANCE;
+    }
+
+    /** Ordered list of URI converters */
+    private final List<UriConverter> uriConverters;
+
+    /**
+     * Creates the URI converter manager and initializes the converter chain.
+     *
+     * <p>
+     * The order is significant:
+     * <ol>
+     *   <li>{@link JarUriConverter} – JAR and JRT resources</li>
+     *   <li>{@link WslUriConverter} – WSL UNC paths</li>
+     *   <li>{@link DefaultUriConverter} – standard file system</li>
+     * </ol>
+     * </p>
+     */
+    UriConverterManager() {
+        uriConverters = List.of(
+                new JarUriConverter(),
+                new WslUriConverter(),
+                new DefaultUriConverter()
+        );
+    }
+
+    /**
+     * Resolves a {@link VirtualFile} from a URI string.
+     *
+     * <p>
+     * Some language servers send percent-encoded URIs that IntelliJ cannot
+     * resolve directly (e.g. Windows drive letters such as {@code c%3A}).
+     * </p>
+     *
+     * <p>
+     * To work around this:
+     * <ul>
+     *   <li>{@code +} characters are preserved</li>
+     *   <li>the URI is URL-decoded before resolution</li>
+     * </ul>
+     * </p>
+     *
+     * @param uri the URI string
+     * @return the resolved {@link VirtualFile}, or {@code null} if no converter
+     *         can handle the URI
+     */
+    @Override
+    public @Nullable VirtualFile findResourceFor(@NotNull String uri) {
+        if (uri.contains("%")) {
+            // Example:
+            // file:///c%3A/Users/azerr/IdeaProjects/untitled7/test.js
+            //
+            // IntelliJ cannot resolve percent-encoded drive letters,
+            // so the URI must be decoded first.
+
+            // Preserve '+' characters during decoding
+            uri = uri.replace("+", "%2B");
+
+            // Decode percent-encoded characters
+            uri = URLDecoder.decode(uri, StandardCharsets.UTF_8);
+        }
+
+        for (UriConverter uriConverter : uriConverters) {
+            VirtualFile file = uriConverter.findResourceFor(uri);
+            if (file != null) {
+                return file;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} into a {@link URI}.
+     *
+     * <p>
+     * The first converter in the chain capable of handling the file
+     * is used.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the corresponding {@link URI}, or {@code null} if unsupported
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull VirtualFile file) {
+        for (UriConverter uriConverter : uriConverters) {
+            URI uri = uriConverter.toUri(file);
+            if (uri != null) {
+                return uri;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts a Java {@link File} into a {@link URI}.
+     *
+     * <p>
+     * This is primarily used when sending file references to a language server.
+     * </p>
+     *
+     * @param file the Java file
+     * @return the corresponding {@link URI}, or {@code null} if unsupported
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull File file) {
+        for (UriConverter uriConverter : uriConverters) {
+            URI uri = uriConverter.toUri(file);
+            if (uri != null) {
+                return uri;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} into its URI string representation.
+     *
+     * <p>
+     * This is typically used when serializing URIs in LSP messages.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return the URI string, or {@code null} if unsupported
+     */
+    @Override
+    public @Nullable String toString(@NotNull VirtualFile file) {
+        for (UriConverter uriConverter : uriConverters) {
+            String uri = uriConverter.toString(file);
+            if (uri != null) {
+                return uri;
+            }
+        }
+        return null;
+    }
+}

--- a/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/WslUriConverter.java
+++ b/src/main/java/com/redhat/devtools/lsp4ij/internal/uri/WslUriConverter.java
@@ -1,0 +1,110 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.internal.uri;
+
+import com.intellij.openapi.vfs.VirtualFile;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.net.URI;
+
+/**
+ * {@link UriConverter} implementation dedicated to Windows Subsystem for Linux (WSL)
+ * paths exposed as Windows UNC paths.
+ *
+ * <p>
+ * WSL file systems are typically accessible on Windows using UNC paths such as:
+ * <pre>
+ * \\wsl$\Ubuntu\home\\user\project\file.txt
+ * </pre>
+ *
+ * <p>
+ * Language servers expect these paths to be represented as valid {@code file://}
+ * URIs with an authority component.
+ * </p>
+ */
+class WslUriConverter implements UriConverter {
+
+    /**
+     * Resolving a {@link VirtualFile} from a WSL URI is not currently supported.
+     *
+     * <p>
+     * IntelliJ exposes WSL files through dedicated file systems rather than
+     * plain {@code file://} URIs, making reverse resolution non-trivial.
+     * </p>
+     *
+     * @param uri the URI string
+     * @return always {@code null}
+     */
+    @Override
+    public @Nullable VirtualFile findResourceFor(@NotNull String uri) {
+        return null;
+    }
+
+    /**
+     * Converts a Windows UNC path pointing to a WSL file into a {@code file://} URI.
+     *
+     * <p>
+     * Example conversion:
+     * <pre>
+     * \\wsl$\Ubuntu\home\\user\file.txt
+     * →
+     * file://wsl$/Ubuntu/home/user/file.txt
+     * </pre>
+     *
+     * <p>
+     * A lenient URI construction is used to support the non-standard but commonly
+     * used {@code wsl$} authority.
+     * </p>
+     *
+     * @param file the Java file representing a WSL UNC path
+     * @return the corresponding {@link URI}, or {@code null} if the file
+     *         is not a WSL UNC path
+     */
+    @Override
+    public @Nullable URI toUri(@NotNull File file) {
+        String path = file.getPath();
+        if (path.startsWith("\\\\")) {
+            // Strip leading UNC prefix (\\)
+            String uncPath = path.substring(2);
+
+            // Extract authority and remaining path: authority\path\...
+            int firstSep = uncPath.indexOf('\\');
+            if (firstSep > 0) {
+                String authority = uncPath.substring(0, firstSep);
+                String uriPath = uncPath
+                        .substring(firstSep)
+                        .replace('\\', '/');
+
+                // Use lenient parsing to allow WSL authority such as "wsl$"
+                return URI.create("file://" + authority + uriPath);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts an IntelliJ {@link VirtualFile} to a URI string.
+     *
+     * <p>
+     * This operation is not currently supported for WSL virtual files and is
+     * expected to be handled by a higher-level converter if needed.
+     * </p>
+     *
+     * @param file the IntelliJ virtual file
+     * @return always {@code null}
+     */
+    @Override
+    public @Nullable String toString(@NotNull VirtualFile file) {
+        return null;
+    }
+}

